### ISSUE TITLE
cache has TypeError with JWT authentication

### ISSUE THREAD

**Issue by leedongwei** [feathers-plus/feathers-hooks-common#327]
### Steps to reproduce
1. I have created an [example repo](https://github.com/leedongwei/feathers-hooks-common-cache-issue), please download it
2. `npm install` on the repo to install dependencies
3. `npm start` to run the app
4. Go to `localhost:3030` on your browser
    * I have added shortcut buttons to login/logout
    * Code is at `script.js`, it will log stuff to the browser console
5. Click `Create user`
6. Click `Login user`
7. Refresh the page to trigger authentication with JWT
8. Looking at the terminal, you should see this
```
info: Feathers application started on http://localhost:3030
info: after: users - Method: find
info: error: authentication - Method: create: Invalid login
error:  NotAuthenticated: Invalid login
    at new NotAuthenticated (/cache-test/node_modules/@feathersjs/errors/lib/index.js:88:17)
    at app.authenticate.then (/cache-test/node_modules/@feathersjs/authentication/lib/hooks/authenticate.js:76:31)
    at <anonymous>
info: after: users - Method: create
info: after: users - Method: find
info: after: authentication - Method: create
info: after: authentication - Method: remove
info: error: users - Method: get: Cannot read property 'id' of undefined
error:  TypeError: Cannot read property 'id' of undefined
    at Object.context (/cache-test/node_modules/feathers-hooks-common/lib/services/cache.js:31:40)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
info: after: authentication - Method: create
```
9. Go to `users.hooks.js`, comment out the lines with `cache(cacheMapUser),`.
    * Line 14 and 26
10. Restart the app
11. Repeat steps 5-8, and there will be no errors

### Expected behavior
* No TypeError

### Actual behavior
* TypeError with cache

### System configuration
* Not applicable

**Comment by eddyystop** [comment]
I will look into this when I return from vacation next Wednesday.

**Comment by eddyystop** [comment]
I get the following on `npm start`

```text

added 608 packages in 119.135s
feathers-hooks-common-cache-issue$ npm start

> cache-test@0.0.0 start /home/johnsz/shared/debugging/feathers-hooks-common-cache-issue
> node src/

sequelize deprecated String based operators are now deprecated. Please use Symbol based operators for better security, read more at http://docs.sequelizejs.com/manual/tutorial/querying.html#operators node_modules/sequelize/lib/sequelize.js:236:13
info: Feathers application started on http://localhost:3030
{ NotFound: Page not found
    at NotFound (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/@feathersjs/errors/lib/index.js:109:17)
    at /home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/@feathersjs/errors/lib/not-found-handler.js:5:10
    at Layer.handle [as handle_request] (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:317:13)
    at /home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:335:12)
    at next (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:275:10)
    at exposeCookies (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/@feathersjs/authentication/lib/express/expose-cookies.js:10:5)
    at Layer.handle [as handle_request] (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:317:13)
    at /home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:335:12)
    at next (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:275:10)
    at exposeHeaders (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/@feathersjs/authentication/lib/express/expose-headers.js:10:5)
    at Layer.handle [as handle_request] (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:317:13)
    at /home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:335:12)
    at next (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:275:10)
    at /home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/@feathersjs/express/lib/rest/index.js:34:7
    at Layer.handle [as handle_request] (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/express/lib/router/index.js:317:13)
  type: 'FeathersError',
  name: 'NotFound',
  message: 'Page not found',
  code: 404,
  className: 'not-found',
  data: undefined,
  errors: {} }
{ NotFound: Page not found
    at NotFound (/home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/@feathersjs/errors/lib/index.js:109:17)
    at /home/johnsz/shared/debugging/feathers-hooks-common-cache-issue/node_modules/@feathersjs/errors/lib/not-found-handler.js:5:10

```

**Comment by leedongwei** [comment]
Thanks for looking into this issue. 

What's your Ubuntu VM? I'll install it and try to debug from there.

**Comment by eddyystop** [comment]
That's kind of you. I'm using Oracle VM VirtualBox 16.04 LTS 64 bits.

**Comment by leedongwei** [comment]
I caught the flu and I've been sleeping it off. I'll get to this sometime later this week.

**Comment by eddyystop** [comment]
Influenza is not a sickness to take lightly. I hope you get well soon.

**Comment by gerwintown** [comment]
I had the exact same issue with the GET method.  After reading the docs and some digging into cache.js, I think I figured out the issue.

In the cache hook, you are using the getItems utility hook to grab the record, which is looking for "context.data," but per the docs:
#### context.id
>context.id is a writeable property and the id for a **get**, remove, update and patch service method >call. For remove, update and patch context.id can also be null when modifying multiple entries. In >all other cases it will be undefined. 

So items on a GET were returning undefined because there is only context.id, not context.data.

This what I did to temporarily solve this problem:

`let items = getItems(context) || {id : context.id};`

Just thought I would share for now, its late here so tomorrow I will probably try to place this solution in my own code and copy context.id to context.data or something.

**Comment by eddyystop** [comment]
Thanks for identifying the problem. I think a cleaner fix is changing
```js
      case 'get':
        const item = items[0];
        const idName = getIdName(keyFieldName, item);
        const value = cacheMap.get(item[idName]);
```
to
```js
      case 'get':
        const value = cacheMap.get(context.id);
```

**Comment by eddyystop** [comment]
Fixed via https://github.com/feathers-plus/feathers-hooks-common/pull/352

Published to npm as 4.5.3.
